### ISSUE TITLE
Possible memory leak?

### ISSUE THREAD

**Issue by evcu** [tensorflow/tensorflow#21584]
Hi I was looking into this. The fact that we are adding the loss to the total inside the gradient_tape would save all the previous iterations within the epoch. Am I wrong?

https://github.com/tensorflow/tensorflow/blob/2942555be63f2dbf557cdf7be2bcbd5d4e9d0daf/tensorflow/contrib/eager/python/examples/gan/mnist.py#L243

**Comment by asimshankar** [comment]
This question is better asked on [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a bug or feature request. There is also a larger community that reads questions there.

If you think we've misinterpreted a bug, please comment again with a clear explanation, as well as all of the information requested in the [issue template](https://github.com/tensorflow/tensorflow/issues/new). Thanks!

That said, Short answer: Nope, this should not be a leak.

Long answer: `total_discriminator_loss += discriminator_loss_val` is shorthand for `total_discriminator_loss = tf.add(total_discriminator_loss, discriminator_loss_val)`. So, the tape only records the `add` operation, furthermore the tape is cleared on each call to `gradient` ([doc](https://www.tensorflow.org/api_docs/python/tf/GradientTape#class_gradienttape)).

Hope that helps.
